File: client-sdk/python/voice-dialogue-agent/event_output.py
from __future__ import annotations

import json


def format_command_event(event: object) -> str | None:
    event_type = getattr(event, "type", "")
    if event_type == "command.called":
        arguments = json.dumps(
            getattr(event, "arguments", {}),
            ensure_ascii=False,
            sort_keys=True,
            default=dict,
        )
        return (
            "command.called: "
            f"{getattr(event, 'command_name', '')} "
            f"call_id={getattr(event, 'call_id', '')} arguments={arguments}"
        )
    if event_type == "command.completed":
        return (
            "command.completed: "
            f"{getattr(event, 'command_name', '')} "
            f"message={getattr(event, 'message', None)!r}"
        )
    if event_type == "command.failed":
        return (
            "command.failed: "
            f"{getattr(event, 'command_name', '')} "
            f"message={getattr(event, 'message', '')!r}"
        )
    return None


def format_turn_latency(event: object) -> str:
    latency = getattr(event, "latency", None)
    stages = getattr(latency, "stages", None)
    measured = (
        ("vad", getattr(stages, "vad_ms", None)),
        ("asr", getattr(stages, "asr_ms", None)),
        ("llm_first", getattr(stages, "llm_first_token_ms", None)),
        ("tts_first", getattr(stages, "tts_first_audio_ms", None)),
        ("playback", getattr(stages, "playback_ms", None)),
        ("total", getattr(latency, "total_ms", None)),
    )
    parts = [f"{name}={value}ms" for name, value in measured if value is not None]
    return "turn.latency" + (f": {' '.join(parts)}" if parts else "")


def show_event(event: object) -> None:
    event_type = getattr(event, "type", "")
    if event_type == "error":
        error = getattr(event, "error", None)
        to_public_dict = getattr(error, "to_public_dict", None)
        details = (
            dict(to_public_dict())
            if callable(to_public_dict)
            else {"source": "sdk", "message": "Unknown SDK error"}
        )
        print(f"error: {json.dumps(details, ensure_ascii=False, sort_keys=True)}", flush=True)
        return
    if event_type == "interruption":
        print(
            "interruption: "
            f"reason={getattr(event, 'reason', '')} "
            f"turn_id={getattr(event, 'turn_id', '')}",
            flush=True,
        )
        return
    if event_type == "image.captured":
        print(
            "image.captured: "
            f"mime={getattr(event, 'mime_type', '')} "
            f"size={getattr(event, 'width', 0)}x{getattr(event, 'height', 0)} "
            f"bytes={getattr(event, 'size_bytes', 0)} "
            f"capture_ms={getattr(event, 'capture_ms', 0)}",
            flush=True,
        )
        return
    command_line = format_command_event(event)
    if command_line is not None:
        print(command_line, flush=True)
        return
    if event_type == "reply.partial":
        print(getattr(event, "text", ""), end="", flush=False)
        return
    if event_type == "turn.latency":
        print(f"\n{format_turn_latency(event)}", flush=True)
        return
    if event_type == "emotion.detected":
        confidence = getattr(event, "confidence", None)
        suffix = (
            f" confidence={confidence:.3f}"
            if isinstance(confidence, (int, float))
            else ""
        )
        print(
            "\nemotion.detected: "
            f"{getattr(event, 'emotion_code', '')} "
            f"source={getattr(event, 'source', '')}{suffix}",
            flush=True,
        )
        return
    if event_type in {
        "speech.started",
        "speech.stopped",
        "transcript.partial",
        "transcript.final",
        "reply.started",
        "reply.final",
        "playback.started",
        "playback.stopped",
    }:
        text = getattr(event, "text", "")
        confidence = getattr(event, "metadata", {}).get("confidence")
        suffix = (
            f" confidence={confidence:.3f}"
            if event_type == "speech.started" and isinstance(confidence, (int, float))
            else ""
        )
        print(f"{event_type}: {text}{suffix}", flush=True)

File: client-sdk/cpp/voice-dialogue-agent/src/event_output.cpp
#include "event_output.hpp"

#include <iostream>
#include <type_traits>
#include <variant>

namespace {
const char *media_role_name(eva::MediaRole role) {
  switch (role) {
  case eva::MediaRole::audio_input:
    return "audio_input";
  case eva::MediaRole::audio_output:
    return "audio_output";
  case eva::MediaRole::aec:
    return "aec";
  case eva::MediaRole::camera:
    return "camera";
  }
  return "unknown";
}

const char *media_operation_name(eva::MediaOperation operation) {
  switch (operation) {
  case eva::MediaOperation::start:
    return "start";
  case eva::MediaOperation::frames:
    return "frames";
  case eva::MediaOperation::stop:
    return "stop";
  case eva::MediaOperation::enqueue:
    return "enqueue";
  case eva::MediaOperation::flush:
    return "flush";
  case eva::MediaOperation::drain:
    return "drain";
  case eva::MediaOperation::process_near_end:
    return "process_near_end";
  case eva::MediaOperation::push_far_end:
    return "push_far_end";
  case eva::MediaOperation::reset:
    return "reset";
  case eva::MediaOperation::release:
    return "release";
  case eva::MediaOperation::capture:
    return "capture";
  }
  return "unknown";
}

const char *media_reason_name(eva::MediaReason reason) {
  switch (reason) {
  case eva::MediaReason::not_configured:
    return "not_configured";
  case eva::MediaReason::permission_denied:
    return "permission_denied";
  case eva::MediaReason::device_unavailable:
    return "device_unavailable";
  case eva::MediaReason::unsupported:
    return "unsupported";
  case eva::MediaReason::timeout:
    return "timeout";
  case eva::MediaReason::invalid_data:
    return "invalid_data";
  case eva::MediaReason::operation_failed:
    return "operation_failed";
  }
  return "unknown";
}

template <typename T>
void print_optional_ms(const char *name, const std::optional<T> &value) {
  if (value)
    std::cout << ' ' << name << '=' << *value << "ms";
}
} // namespace

void show_error(std::ostream &output, std::string_view action,
                const eva::Error &error) {
  output << "error: action=" << action
         << " source=" << eva::error_source_name(error.source())
         << " fatal=" << (error.fatal() ? "true" : "false")
         << " message=" << error.message();
  if (error.provider())
    output << " provider=" << *error.provider();
  if (error.status_code())
    output << " status_code=" << *error.status_code();
  if (error.role())
    output << " media_role=" << media_role_name(*error.role());
  if (error.operation())
    output << " media_operation=" << media_operation_name(*error.operation());
  if (error.reason())
    output << " media_reason=" << media_reason_name(*error.reason());
  if (error.trace_id())
    output << " trace_id=" << *error.trace_id();
  output << '\n';
}

void show_event(const eva::AgentEvent &event) {
  std::visit(
      [](const auto &value) {
        using T = std::decay_t<decltype(value)>;
        if constexpr (std::is_same_v<T, eva::SpeechStartedEvent>) {
          std::cout << "speech.started\n";
        } else if constexpr (std::is_same_v<T, eva::SpeechStoppedEvent>) {
          std::cout << "speech.stopped\n";
        } else if constexpr (std::is_same_v<T, eva::TranscriptPartialEvent>) {
          std::cout << "transcript.partial: " << value.text << '\n';
        } else if constexpr (std::is_same_v<T, eva::TranscriptFinalEvent>) {
          std::cout << "transcript.final: " << value.text << '\n';
        } else if constexpr (std::is_same_v<T, eva::ReplyStartedEvent>) {
          std::cout << "reply.started\n";
        } else if constexpr (std::is_same_v<T, eva::ReplyPartialEvent>) {
          std::cout << "reply.partial: " << value.text << '\n';
        } else if constexpr (std::is_same_v<T, eva::ReplyFinalEvent>) {
          std::cout << "reply.final: " << value.text << '\n';
        } else if constexpr (std::is_same_v<T, eva::PlaybackStartedEvent>) {
          std::cout << "playback.started\n";
        } else if constexpr (std::is_same_v<T, eva::PlaybackStoppedEvent>) {
          std::cout << "playback.stopped\n";
        } else if constexpr (std::is_same_v<T, eva::EmotionDetectedEvent>) {
          std::cout << "emotion.detected: code=" << value.emotion_code
                    << " latency=" << value.latency_ms << "ms";
          if (value.confidence)
            std::cout << " confidence=" << *value.confidence;
          std::cout << '\n';
        } else if constexpr (std::is_same_v<T, eva::ImageCapturedEvent>) {
          std::cout << "image.captured: mime=" << value.mime_type
                    << " size=" << value.width << 'x' << value.height
                    << " bytes=" << value.size_bytes
                    << " capture=" << value.capture_ms << "ms\n";
        } else if constexpr (std::is_same_v<T, eva::InterruptionEvent>) {
          std::cout << "interruption: turn_id=" << value.turn_id << '\n';
        } else if constexpr (std::is_same_v<T, eva::CommandCalledEvent>) {
          std::cout << "command.called: " << value.command_name
                    << " call_id=" << value.call_id << '\n';
        } else if constexpr (std::is_same_v<T, eva::CommandCompletedEvent>) {
          std::cout << "command.completed: " << value.command_name
                    << " call_id=" << value.call_id;
          if (value.message)
            std::cout << " message=" << *value.message;
          std::cout << '\n';
        } else if constexpr (std::is_same_v<T, eva::CommandFailedEvent>) {
          std::cout << "command.failed: " << value.command_name
                    << " call_id=" << value.call_id
                    << " message=" << value.message << '\n';
        } else if constexpr (std::is_same_v<T, eva::TurnLatencyEvent>) {
          std::cout << "turn.latency:";
          print_optional_ms("vad", value.latency.stages.vad_ms);
          print_optional_ms("asr", value.latency.stages.asr_ms);
          print_optional_ms("llm_first",
                            value.latency.stages.llm_first_token_ms);
          print_optional_ms("tts_first",
                            value.latency.stages.tts_first_audio_ms);
          print_optional_ms("playback", value.latency.stages.playback_ms);
          print_optional_ms("total", value.latency.total_ms);
          std::cout << '\n';
        } else if constexpr (std::is_same_v<T, eva::ErrorEvent>) {
          show_error(std::cerr, "event", value.error);
        }
      },
      event);
}
